language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -315,3 +362,34 @@ onHead :: (a -> a) -> [a] -> [a]
 onHead f = \case
     [] -> []
     x:xs -> f x:xs
+
+onLast :: (a -> a) -> [a] -> [a]
+onLast f = \case
+  [] -> []
+  [l] -> [f l]
+  x : xs -> x : onLast f xs
+
+-- A type such as
+--   data Q = C { x: Int, y: Text }| G { z: Bool }
+-- has a DAML-LF representation like,
+--   record Q.C = { x: Int, y: String }
+--   record Q.G = { z: Bool }
+--   variant Q = C Q.C | G Q.G
+-- Use this function to get the definitions of 'Q.C' and 'Q.G' given
+-- 'Q'.
+assocDefDataTypes :: Module -> T.Text -> [((T.Text, T.Text), DefDataType)]
+assocDefDataTypes m c =
+  [((sup, sub ), def) | def <- defDataTypes m
```

Comment:
```suggestion
  [(sub, def) | def <- defDataTypes m
```
This feels quite "unnormalized" to me.

Reply:
Not sure what you are getting at. There are two uses
```
                  assocSers = map
                    (\(n, d) -> serFromDef id n curPkgId mod tpls d)
                    assocDefs
```
and
```
                  assocSers = map
                    (\(n@(_, c), d) -> (c, serFromDef (drop 1) n curPkgId mod tpls d))
                    assocDefs
```
I designed it together with `serFromDef` to play nice.

Reply:
The very first element of each list item is the same, namely the second parameter `c`. I suggest we don't return this duplicated information but rather change the call site to explicitly use whatever we passed as the second argument to `assocDefDataTypes` instead.

Reply:
Good reason. Fixed.